src/apex.js

```diff
@@ -109,3 +130,53 @@ async function onInbox ({ actor, activity, recipient, object }) {
     return apex.publishUpdate(recipient, await apex.getFollowers(recipient))
   }
 }
+
+// complex scoping by activity type for outbox post
+const profileUpdateProps = ['id', 'name', 'icon', 'avatar', 'summary']
+function outboxScoping (req, res, next) {
+  const authorizedScope = req.authInfo?.scope || []
+  let postType = req.body?.type?.toLowerCase?.()
+  const object = req.body?.object?.[0]
+  if (
+    postType === 'update' &&
+    // update target is the actor itself
+    object?.id === res.locals.apex.target?.id &&
+    Object.keys(object).every(prop => profileUpdateProps.includes(prop))
+  ) {
+    // profile udpates are lower permission than general update
+    postType = 'update-profile'
+  }
+  // default requires unlimited access
+  const requiredScope = ['*']
+  switch (postType) {
+    case 'arrive':
+    case 'leave':
+      requiredScope.push(scopes.postLocation.name)
+      break
+    case 'follow':
+    case 'accept':
+      requiredScope.push(scopes.addFriends.name)
+      break
+    case 'block':
+      requiredScope.push(scopes.addBlocks.name)
+      break
+    case 'add':
+    case 'create':
+    case 'like':
+    case 'announce':
+    case 'update-profile':
+      requiredScope.push(scopes.creative.name)
+      break
+    case 'update':
+    case 'reject':
+    case 'undo':
+    case 'delete':
+    case 'remove':
+      requiredScope.push(scopes.destructive.name)
+      break
+  }
+  if (!overlaps(requiredScope, authorizedScope)) {
+    res.locals.apex.authorized = false
```

Comment:
All we have to do to block access on a specific request is set this flag and `activitypub-express` handles the rest